drivers/input/input_cf1133.c

```diff
@@ -0,0 +1,326 @@
+
+/* cf1133 used registers */
+#define FIRMWARE_VERSION    0x0
+#define STATUS_REG          0x1
+#define DEVICE_CONTROL_REG  0x2
+#define TIMEOUT_TO_IDLE_REG 0x3
+#define XY_RESOLUTION_HIGH  0x4
+#define X_RESOLUTION_LOW    0x5
+#define Y_RESOLUTION_LOW    0x6
+#define DEVICE_CONTROL_REG2 0x09
+#define FIRMWARE_REVISION_3 0x0C
+#define FIRMWARE_REVISION_2 0x0D
+#define FIRMWARE_REVISION_1 0x0E
+#define FIRMWARE_REVISION_0 0x0F
+#define FINGERS             0x10
+#define KEYS_REG            0x11
+#define XY0_COORD_H         0x12
+#define X0_COORD_L          0x13
+#define Y0_COORD_L          0x14
+#define I2C_PROTOCOL        0x3E
+#define MAX_NUM_TOUCHES     0x3F
+#define DATA_0_HIGH         0x40
+#define DATA_0_LOW          0x41
+#define MISC_CONTROL        0xF1
+#define SMART_WAKE_UP_REG   0xF2
+#define CHIP_ID             0xF4
+#define PAGE_REG            0xFF
+
+/* Constants */
+#define SUPPORTED_POINT			 0x1
+#define PIXEL_DATA_LENGTH_B      0x3
+#define PIXEL_DATA_LENGTH_A      0x4
+#define SITRONIX_RESERVED_TYPE_0 0x0
+#define SITRONIX_A_TYPE          0x1
+#define SITRONIX_B_TYPE          0x2
+
+/* Mask */
+#define ONE_D_SENSING_CONTROL_SHFT GENMASK(1, 1)
+#define ONE_D_SENSING_CONTROL_BMSK GENMASK(1, 0)
+#define I2C_PROTOCOL_BMSK          GENMASK(1, 0)
+#define TOUCH_POINT_VALID_MSK	   GENMASK(7, 7)
+
+/* Offset for coordinates registers */
+#define XY_COORD_H               0x0
+#define X_COORD_L                0x1
+#define Y_COORD_L                0x2
+
+/* CF1133 configuration (DT) */
+struct cf1133_config {
+	/** I2C bus. */
+	struct i2c_dt_spec bus;
+#ifdef CONFIG_INPUT_CF1133_INTERRUPT
+	/** Interrupt GPIO information. */
+	struct gpio_dt_spec int_gpio;
+#endif
+};
+
+/* CF1133 data */
+struct cf1133_data {
+	/** Device pointer. */
+	const struct device *dev;
+	/** Work queue (for deferred read). */
+	struct k_work work;
+#ifdef CONFIG_INPUT_CF1133_INTERRUPT
+	/** Interrupt GPIO callback. */
+	struct gpio_callback int_gpio_cb;
+#else
+	/* Timer (polling mode) */
+	struct k_timer timer;
+#endif
+	/* Last pressed state */
+	bool pressed_old;
+	bool pressed;
+
+	int resolution_x;
+	int resolution_y;
+	uint8_t touch_protocol_type;
+	uint8_t pixel_length;
+	uint8_t chip_id;
+};
+
+static int cf1133_get_chip_id(struct cf1133_data *data, const struct cf1133_config *config)
+{
+	int ret;
+	uint8_t buffer[3];
+	uint8_t num_x;
+	uint8_t num_y;
+
+	ret = i2c_burst_read_dt(&config->bus, CHIP_ID, buffer, sizeof(buffer));
+	if (ret < 0) {
+		LOG_ERR("Read burst failed: %d", ret);
+		return ret;
+	}
+	if (buffer[0] == 0) {
+		if (buffer[1] + buffer[2] > 32) {
+			data->chip_id = 2;
+		} else {
+			data->chip_id = 0;
+		}
+	} else {
+		data->chip_id = buffer[0];
+	}
+	num_x = buffer[1];
+	num_y = buffer[2];
+	LOG_DBG("Chip ID = %d, num_x = %d, num_y = %d", data->chip_id, num_x, num_y);
+
+	return 0;
+}
+
+static int cf1133_get_protocol_type(struct cf1133_data *data, const struct cf1133_config *config)
```

Comment:
Maybe its a preference thing, but there are some methods that take the deconstructed data and config and some take the entire device.

Reply:
@clemdy I would recommend just passing `dev` to all internal functions and then get data and config internally, there's no performance penalties (the compiler optimizes all the pointer arithmetic out I think), you save a tiny bit of stack (only on pointer to pass) and you don't have to update the function signature if you need more stuff.